plat/*: Add IRQ events

jschlumpp opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64`, `arm`
 - Platform(s): `kvm`, `xen`, `linuxu`
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

This allows external libraries to hook into any IRQ by using the ukevent functionality. An event handler can also decide whether to pass the control flow to other event handlers and normal IRQ handlers.

jschlumpp committed:
plat/*: Expose registers to main irq handler function

The trap handlers already pass on the registers to the handler
functions. This passes them to the KVM-internal dispatch function.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

jschlumpp committed:
plat/*: Add event for IRQs

Adds an event which is triggered when an IRQ occurs. The event handler
has access to the IRQ number and the registers of the interrupted
thread.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

mogasergiu commented on plat/kvm/irq.c:
```diff
@@ -86,13 +89,24 @@ int ukplat_irq_register(unsigned long irq, irq_handler_func_t func, void *arg)
+	ctx.regs = regs;
+	ctx.irq = irq;
+	rc = uk_raise_event(UKARCH_EVENT_IRQ, &ctx);
+	if (rc == UK_EVENT_HANDLED)
```
Since you're not using `rc` anywhere else, why not remove it and just
```
if (uk_raise_event(UKARCH_EVENT_IRQ, &ctx) == UK_EVENT_HANDLED)
```

marcrittinghaus replied:
Personally, I prefer the separate `rc` version just for consistency with the rest of the code, but   add a `likely` for good practice. And with the `likely` also the line becomes a bit long and hard to read.

jschlumpp replied:
I don't see a reason to use either `likely` or `unlikely` here. The return values really depends on the event handlers and the parameters. (But there is still the error case `rc < 0` missing, which is introduced in the other PR, for that unlikely makes sense).

For the comments on the `rc` variable I agree with Marc.

mogasergiu commented on plat/linuxu/irq.c:
```diff
@@ -150,9 +153,17 @@ static void _irq_handle(int irq)
 {
 	struct irq_handler *h;
 	int i;
+	int rc;
```
ditto

mogasergiu commented on plat/xen/events.c:
```diff
@@ -85,6 +89,8 @@ void unbind_all_ports(void)
 int do_event(evtchn_port_t port, struct __regs *regs)
 {
 	ev_action_t *action;
+	int rc;
```
ditto

marcrittinghaus commented on plat/drivers/gic/gic-v3.c:
```diff
@@ -499,7 +499,7 @@ static void gicv3_init_dist(void)
 }
 
-static void gicv3_handle_irq(void)
+static void gicv3_handle_irq(struct __regs *regs)
```
Don't forget to add `__unused` attributes if necessary.

jschlumpp replied:
Is this a more general comment? `regs` is used in the function.

marcrittinghaus replied:
Ah, yes. My bad 👍🏻

jschlumpp commented:
I just wanted to adjust this PR to #691 regarding error from event handlers. I only see two options to handle these errors in the interrupt handler: 

1. Call `UK_CRASH`. That means any error returned from an event handler is fatal.
2. Ignore the error and insert a tracepoint for debugging.

(we can't print in the context of the interrupt handler, as the print functionality is not ISR-safe)

I'm currently leaning towards implementing option (1). Are there any opinions or better ideas about this?

marcrittinghaus commented:
As unhandled traps also cause a `UK_CRASH`, it makes sense to go for option 1.

jschlumpp committed:
plat/*: Handle errors returned from IRQ event handlers

PR #691 adds the possibility to return error codes from the event
handler. In the context of interrupts, we interpret an error code as a
fatal error and emit a crash.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

mogasergiu commented on plat/common/arm/traps_arm64.c:
```diff
@@ -208,5 +208,5 @@ void trap_el1_irq(void)
 #endif /* CONFIG_ARM64_FEAT_MTE */
 
-	gic->ops.handle_irq();
+	gic->ops.handle_irq(regs);
```
I believe you wanted to pass `regs` as an argument to `trap_el1_irq` as well?

marcrittinghaus replied:
+1. This will also require changes to the ARM assembly for trap handling just like with x86.

marcrittinghaus commented on include/uk/plat/irq.h:
```diff
@@ -56,6 +58,13 @@ typedef int (*irq_handler_func_t)(void *);
+	unsigned long irq;
+};
+
+#define UKARCH_EVENT_IRQ event_irq
```
We should call it `UKPLAT_` since the ctx structure is also `ukplat` and it is accessible via the `plat` header. Putting in it the `arch/x86/x86_64/include/uk/asm/lcpu.h` header (and duplicate for other architectures) would also work with `ukarch` prefix. However, we are currently discussing that the `arch` folder should only contain things actually defined by the architectural manual. And the events are more of a platform construct anyways.

In any case also put the prefix in the actual event name. We should have done this also for the trap events. I'll probably change it during re-arch. 

Also add a remark when this event is raised.
```suggestion
/** Bla */
#define UKPLAT_EVENT_IRQ ukplat_event_irq
```

marcrittinghaus commented on include/uk/plat/irq.h:
```diff
@@ -56,6 +58,13 @@ typedef int (*irq_handler_func_t)(void *);
+struct ukplat_event_irq_data {
+	struct __regs *regs;
+	unsigned long irq;
+};
```
Please add comments to the structure.
```suggestion
/** Bla */
struct ukplat_event_irq_data {
	/** Bla */
	struct __regs *regs;
	/** Bla */
	unsigned long irq;
};
```

marcrittinghaus commented on plat/kvm/irq.c:
```diff
@@ -86,13 +89,26 @@ int ukplat_irq_register(unsigned long irq, irq_handler_func_t func, void *arg)
+		/* Skip all normal handlers if an event handler handled the
+		 * event
+		 */
+		goto exit_ack;
```
Put `{}` around the block due to the comment. Checkpatch will not complain.

marcrittinghaus commented on plat/common/include/uk/plat/common/irq.h:
```diff
@@ -65,6 +65,6 @@ enum uk_irq_polarity {
  *
  * @param irq IRQ vector
```
```suggestion
 * @param regs ...
 * @param irq IRQ vector
```

marcrittinghaus reviewed (changes requested):
Hey @mschlumpp,

many thanks for this PR! Please see the comments.

marcrittinghaus commented on plat/xen/events.c:
```diff
@@ -38,11 +38,15 @@
 
 #define NR_EVS 1024
 
+UK_EVENT(UKARCH_EVENT_IRQ);
```
```suggestion
UK_EVENT(UKPLAT_EVENT_IRQ);
```

marcrittinghaus commented on plat/xen/events.c:
```diff
@@ -93,6 +99,14 @@ int do_event(evtchn_port_t port, struct __regs *regs)
 
+	ctx.regs = regs;
+	ctx.irq = port;
+	rc = uk_raise_event(UKARCH_EVENT_IRQ, &ctx);
```
```suggestion
	rc = uk_raise_event(UKPLAT_EVENT_IRQ, &ctx);
```

marcrittinghaus commented on plat/linuxu/irq.c:
```diff
@@ -150,9 +153,19 @@ static void _irq_handle(int irq)
 
+	ctx.regs = NULL;
+	ctx.irq = irq;
+	rc = uk_raise_event(UKARCH_EVENT_IRQ, &ctx);
```
```suggestion
	rc = uk_raise_event(UKPLAT_EVENT_IRQ, &ctx);
```

marcrittinghaus commented on plat/linuxu/irq.c:
```diff
@@ -49,6 +50,8 @@
 #define __MAX_IRQ 16
 #define HANDLER_RESERVED ((void *) 0x1)
 
+UK_EVENT(UKARCH_EVENT_IRQ);
```
```suggestion
UK_EVENT(UKPLAT_EVENT_IRQ);
```

marcrittinghaus reviewed (changes requested):
One last thing 😄

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu.moga@protonmail.com>

marcrittinghaus reviewed (approved):
Thanks @mschlumpp, @mogasergiu!

Approved-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>
